pkg/trie/triedb/proof/generate.go

```diff
@@ -0,0 +1,450 @@
+	}
+	e.children[e.childIndex] = childRef
+	e.childIndex++
+}
+
+func (e *stackEntry) completBranchChildren(
+	childHandles [codec.ChildrenCapacity]codec.MerkleValue,
+	childIndex int,
+) {
+	for i := childIndex; i < codec.ChildrenCapacity; i++ {
+		switch n := childHandles[i].(type) {
+		case codec.InlineNode:
+			e.children[i] = triedb.NewInlineChildReference(n.Data)
+		case codec.HashedNode:
+			e.children[i] = triedb.NewHashChildReference(common.Hash(n.Data))
+		}
+	}
+}
+
+func (e *stackEntry) replaceChildRef(encodedChild []byte, child codec.MerkleValue) triedb.ChildReference {
+	switch child.(type) {
+	case codec.HashedNode:
+		return triedb.NewInlineChildReference(nil)
+	case codec.InlineNode:
+		return triedb.NewInlineChildReference(encodedChild)
+	default:
+		panic("unreachable")
+	}
+}
+
+func Generate(db db.RWDatabase, trieVersion trie.TrieLayout, rootHash common.Hash, keys []string) (
+	proof [][]byte, err error) {
+	// Sort and deduplicate keys
+	keys = sortAndDeduplicateKeys(keys)
+
+	// The stack of nodes through a path in the trie.
+	// Each entry is a child node of the preceding entry.
+	stack := deque.New[*stackEntry]()
+
+	// final proof nodes
+	var proofNodes [][]byte
+
+	// Iterate over the keys and build the proof nodes
+	for i := 0; i < len(keys); i = i + 1 {
+		var key = []byte(keys[i])
+		var keyNibbles = nibbles.KeyLEToNibbles(key)
+
+		err := unwindStack(stack, proofNodes, &keyNibbles)
+		if err != nil {
+			return nil, err
+		}
+
+		// Traverse the trie recording the visited nodes
+		recorder := triedb.NewRecorder()
+		trie := triedb.NewTrieDB(rootHash, db, triedb.WithRecorder(recorder))
+		trie.SetVersion(trieVersion)
+		trie.Get(key)
+
+		recordedNodes := triedb.NewRecordedNodesIterator(recorder.Drain())
+
+		// Skip over recorded nodes already on the stack.
+		for i := 0; i < stack.Len(); i++ {
+			nextEntry := stack.At(i)
+			nextRecord := recordedNodes.Peek()
+
+			if nextRecord == nil || !bytes.Equal(*nextEntry.nodeHash, nextRecord.Hash) {
+				break
+			}
+
+			recordedNodes.Next()
+		}
+
+		// Descend in trie collecting nodes until find the value or the end of the path
+	loop:
+		for {
+			var nextStep step
+			var entry *stackEntry
+			if stack.Len() > 0 {
+				entry = stack.Back()
+			}
+			if entry == nil {
+				nextStep = stepDescend{childPrefixLen: 0, child: nodeHandleHash{hash: rootHash.ToBytes()}}
+			} else {
+				var err error
+				nextStep, err = matchKeyToNode(
+					entry.node,
+					&entry.omitValue,
+					&entry.childIndex,
+					keyNibbles,
+					len(entry.prefix),
+					recordedNodes,
+				)
+
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			switch s := nextStep.(type) {
+			case stepDescend:
+				childPrefix := keyNibbles[:s.childPrefixLen]
+				var childEntry *stackEntry
+				switch child := s.child.(type) {
+				case nodeHandleHash:
+					childRecord := recordedNodes.Next()
+
+					if !bytes.Equal(childRecord.Hash, child.hash) {
+						panic("hash mismatch")
+					}
```

Comment:
Do we really need this check? is this one of those `assert_debug` checks?

Reply:
This is an [assert_eq](https://github.com/paritytech/trie/blob/c3348e4b335443c17586a8b7491ba893cfde76a6/trie-db/src/proof/generate.rs#L304) in parity. We can trust and remove it bc this should never happen

Reply:
I'm ok with leaving it in, if they are actually doing this check in the working substrate node.  I'd consider removing it it it was a `debug_assert_eq`.

Reply:
I think if there is no reason to keep it in besides "they do it" then we should probably remove it (this implies being confident in the reason for removal, though there should be confidence in behavior regardless of if the action is remove or keep)

Reply:
but also to be clear, if we do need and its useful than we should keep it